File: LeetCode/1665. Minimum Initial Energy to Finish Tasks/solution.py
class Solution:
    def minimumEffort(self, tasks: List[List[int]]) -> int:
        tasks.sort(key=lambda x: x[0] - x[1])
        p = suma = 0
        for ai, mi in tasks:
            p = max(p, suma + mi)
            suma += ai
        return p


# 作者：zerotrac2
# 链接：https://leetcode-cn.com/problems/minimum-initial-energy-to-finish-tasks/solution/wan-cheng-suo-you-ren-wu-de-zui-shao-chu-shi-neng-/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: LeetCode/1665. Minimum Initial Energy to Finish Tasks/solution.cpp
class Solution {
public:
    int minimumEffort(vector<vector<int>>& tasks) {
        sort(tasks.begin(), tasks.end(), [](vector<int> &a, vector<int> &b){
            return a[1]-a[0] > b[1]-b[0];
        });
        int res = 0, cur = 0;
        for (auto &task : tasks){
            int c = task[0], q = task[1];
            res += max(cur,q)-cur;
            cur = max(cur,q);
            cur -= c;
        }
        return res;
    }
};